### PR TITLE
upstream: log metadata used to create subset lbs

### PR DESCRIPTION
**Pull request by zuercher**
Per a slack conversation, it would be useful is the subset load balancer logged the metadata it used to create it's underlying loadbalancers. This modifies the subset load balancer to track the default subset using the same data type used for non-default subsets (thereby eliminating the `hostMatchesDefaultSubset` function). Logging occurs at the debug level.

Signed-off-by: Stephan Zuercher <stephan@turbinelabs.io>

*Risk Level*: Low

*Testing*: existing unit tests cover the aded code and prove that no functional changes occurred.

### CODE REVIEW

**Comment by junr03** [comment]
@dnoe do you mind taking a first pass?

File: source/common/upstream/subset_lb.cc [hunk 1]
```diff
@@ -124,19 +126,30 @@ SubsetLoadBalancer::LbSubsetEntryPtr SubsetLoadBalancer::findSubset(
...
                                               const HostVector& hosts_removed) {
   if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::NO_FALLBACK) {
+    ENVOY_LOG(debug, "subset lb: fallback load balancer disabled");
     return;
   }
 
   HostPredicate predicate;
 
-  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) {
+  bool fallback_any = (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) ||
```
Comment (mattklein123) [inline source/common/upstream/subset_lb.cc]: nit: const

File: source/common/upstream/subset_lb.cc [hunk 2]
```diff
@@ -273,6 +274,26 @@ SubsetLoadBalancer::extractSubsetMetadata(const std::set<std::string>& subset_ke
   return kvs;
 }
 
+std::string SubsetLoadBalancer::describeMetadata(const SubsetLoadBalancer::SubsetMetadata& kvs) {
```
Comment (mattklein123) [inline source/common/upstream/subset_lb.cc]: Can we test this somehow? I think especially with https://github.com/envoyproxy/envoy/pull/2751 this won't even get run in most normal cases.

Comment (zuercher) [reply source/common/upstream/subset_lb.cc]: Added a test case for this.

File: test/common/upstream/subset_lb_test.cc [hunk 3]
```diff
@@ -1110,6 +1115,50 @@ TEST_P(SubsetLoadBalancerTest, ZoneAwareBalancesSubsetsAfterUpdate) {
...
+  }
+  ASSERT(upstream_logger != nullptr);
+  upstream_logger->setLevel(spdlog::level::debug);
+
+  init({
+      {"tcp://127.0.0.1:80", {{"version", "new"}}},
+      {"tcp://127.0.0.1:81", {{"version", "default"}}},
+  });
+
+  upstream_logger->setLevel(spdlog::level::err);
```
Comment (mattklein123) [inline test/common/upstream/subset_lb_test.cc]: If someone configured a different log level for testing on the CLI with --test_arg, won't this overwrite it after the test runs? I wish there was a better way of doing a test like this, but at minimum should we capture the previous log level so we can reset it?

Comment (mattklein123) [reply test/common/upstream/subset_lb_test.cc]: One option would be to just test the print function directly and not the logging. Up to you.

File: source/common/upstream/subset_lb.cc [hunk 4]
```diff
@@ -232,21 +244,6 @@ void SubsetLoadBalancer::update(uint32_t priority, const HostVector& hosts_added
                  });
 }
 
-bool SubsetLoadBalancer::hostMatchesDefaultSubset(const Host& host) {
```
Comment (mattklein123) [inline source/common/upstream/subset_lb.cc]: Nice cleanup

**Pull request merged**